### PR TITLE
feat: Add support for custom scalar validation

### PR DESCRIPTION
**Pull request by bboure**
This PR fixes the following issues:

- In GraphQL, optional input variables and types are also nullable.
- in the JSON Schema, `enum` should not be combined with `type` (removed)
- custom scalars should only accept basic primitives (i.e. no objects, arrays, etc)

### CODE REVIEW

File: packages/graphql-language-service/src/utils/getVariablesJSONSchema.ts [hunk 1]
```diff
@@ -139,22 +140,48 @@ function getJSONSchemaFromGraphQLType(
...
+    // default scalars
+    if (scalarTypesMap[type.name]) {
+      if (isNonNull) {
+        definition.type = scalarTypesMap[type.name];
+      } else {
+        definition.type = [scalarTypesMap[type.name], 'null'];
+      }
+    } else {
+      // custom scalar can be of any type
+      // FIXME: allow custom scalars to be configured somewhere?
```
Comment (acao) [inline packages/graphql-language-service/src/utils/getVariablesJSONSchema.ts:161]: these are determined from the custom scalar spec https://graphql.org/blog/2023-01-14-graphql-scalars/

so the mapping should be based on the provided scalars in the official repo

**Comment by acao** [comment]
awesome work! can you add a changeset and another test case for the null case that isn't covered? the coverage info you can see on the files tab

File: packages/graphql-language-service/src/utils/getVariablesJSONSchema.ts (hunk 1 above)
Comment (bboure) [reply packages/graphql-language-service/src/utils/getVariablesJSONSchema.ts:161]: Good point, we probably could implement validation for "official" specs.

But one might want custom or non-spec'ed scalars, so I'd still like a way to inject custom validation.
(example, [AWS AppSync scalars.](https://docs.aws.amazon.com/appsync/latest/devguide/scalars.html)). 

Also, one might want to override the community specs (e.g. custom `DateTime` that does not follow the specs for some reason).

I'd be happy to work on that, but it's kinda out of the scope of this PR.

Comment (acao) [reply packages/graphql-language-service/src/utils/getVariablesJSONSchema.ts:161]: indeed, i agree it's out of scope, just adding a meta-comment for your comment haha

**Comment by acao** [comment]
@bboure why don't you go ahead and add the custom mappings in this PR then, I agree it makes sense. 

Others have expressed this need, and even introspecting the scalar info doesn't give us a deterministic solution. perhaps we can even provide a default mapping that can be overridden by users? Or perhaps we should provide none to avoid confusion.

Bonus if you can add the new scalar mapping to monaco-graphql as well, but if not I can. The main thing to remember is that the inter process boundary between main and worker process means the mapping will need to be json serializeable, though you don't need to perform the serialization.

it just means we will need a flat object mapping rather than a callback function or some such

Also, be sure to update the changelog and readme to reflect the changed behaviors with scalars

**Comment by acao** [comment]
by flat mapping, it could look like this:

```js
{ 
  DateTime: { type: "string", format: "datetime" }
}
```

come to think of it, i made a mapping just like this for a proprietary project where we decoratively generated openapi3 specs from graphql schema, I'm embarassed I didn't provide it sooner!

**Comment by bboure** [comment]
@acao  ok, I'll give it a shot.

I'll give a shot at adding it to `monaco-graphql` since this is something that I'll need, but I might need some guidance.

I was thinking in that direction too.

I think one good option would be to add an option somewhere in `DiagnosticSettings`. We'll probably need a config per model uri.

Something like this

```ts
monacoGraphQLAPI.setDiagnosticSettings({
  validateVariablesJSON: {...},
  customScalarSchemaMappings: {
    [Uri.file('operation.graphql').toString()]: {
      DateTime: { type: "string", format: "date-time" },
      // maybe a short-hand for type could be useful but not strictly necessary as it might bring confusion too.
      MyCustomScarlar: 'string',
    }
  }
});
```

**Comment by acao** [comment]
@bboure awesome!  I think this could work fine, but I wonder if configuring it per schema makes even more sense?

```ts
initializeMode({
  schemas: [
    {
      uri: 'https://my-schema.com',
      fileMatch: ['**/*.graphql'],
      // note: not sure if ^ works if the graphql model is using urls for uris?
      customScalarMappings: {
          DateTime: { type: "string", format: "date-time" },
          // maybe a short-hand for type could be useful but not strictly necessary as it might bring confusion too.
          MyCustomScarlar: 'string',
      }
    },
  ],
});
```

**Comment by acao** [comment]
also, in case you're wondering why this lives on `graphql-language-service`, I have plans to use this logic for variables validation/completion in the vscode extension as well :)

**Comment by bboure** [comment]
@acao I added custom scalar schemas support and `monaco-graphql` config as suggested.

I also took the opportunity to improve the schema descriptions a bit.
Descriptions will now look as follow:

```
[field description] (for object types and if field description is present)
(two new lines)
Type (scalar or custom)
[type description] (If present, except for default scalars)
```

Note about type descriptions: I kept backward compatibility and do not render the description of default scalars (String, Int, etc). (They were probably removed for a reason, probably because they are obvious?)

For custom scalars, we use the description from the custom schema, if present.

Examples
```
The birth date of the user.

Date!
An ISO_8601 date
```

```
The name of the user

String!
```

lmk if you have comments

**Comment by acao** [comment]
@bboure nice, and input objects retain their formatting as well?

**Comment by bboure** [comment]
the above applies to all types. Objects might look like this


```
A field description

MyObjectType!
My object description
```

**Comment by acao** [comment]
hm,,. testing this out in vscode and monaco-graphql, it seems that something breaks and the hover no longer renders, I will need to see if something has regressed elsewhere

**Comment by acao** [comment]
another issue with monaco is the schema url, the warning blocks validation messages

can you change the meta schema to draft 4? this is what package.json and many standards use, and is mostly all we need for the simple schemas we generate from variable definitions
`'http://json-schema.org/draft-04/schema'`

**Comment by acao** [comment]
I have some other changes I was planning to add today, for example to add optional anchor links for the graphiql@4 release with monaco coming soon, but I will wait until this PR is complete

**Comment by bboure** [comment]
> update: the issue is with regressions in the example projects from what I can tell

Where is the issue exactly? 

did I break something?

**Comment by bboure** [comment]
I tried this in one of my projects. it seems to work fine

image

**Comment by acao** [comment]
> did I break something?

no, there is a regression in the demos, where someone changed the mode instantiation pattern, so when the schema is fetched, the validation doesn't happen automatically for queries and variables, so you have to modify the query first before variables are validated again

**Comment by bboure** [comment]
> no, there is a regression in the demos, where someone changed the mode instantiation pattern, so when the schema is fetched, the validation doesn't happen automatically for queries and variables, so you have to modify the query first before variables are validated again

Ah! I noticed this too and drove me crazy. It was next on my list to have a look at and fix. 
You'll probably know better than me how to fix, but if you point me in the right direction, I can also have a look

**Comment by acao** [comment]
@bboure I have some more fixes on the way, as you can see in #3384

File: packages/graphql-language-service/src/utils/getVariablesJSONSchema.ts [hunk 2]
```diff
@@ -99,14 +141,14 @@ function renderTypeToString(
...
-  Boolean: 'boolean',
+const defaultScalarTypesMap: { [key: string]: JSONSchema6 } = {
+  Int: { type: 'integer' },
+  String: { type: 'string' },
+  Float: { type: 'number' },
+  ID: { type: 'string' },
+  Boolean: { type: 'boolean' },
   // { "type": "string", "format": "date" } is not compatible with proposed DateTime GraphQL-Scalars.com spec
-  DateTime: 'string',
+  DateTime: { type: 'string' },
```
Comment (acao) [inline packages/graphql-language-service/src/utils/getVariablesJSONSchema.ts:151]: is there a reason we don't specify `format: 'date-time'` here?
I think these are compatible:
https://scalars.graphql.org/andimarek/date-time
https://json-schema.org/draft/2020-12/json-schema-validation.html#section-7.3.1

**Comment by bboure** [comment]
@acao  I can rebase as soon as you merge the other.

File: packages/graphql-language-service/src/utils/getVariablesJSONSchema.ts (hunk 2 above)
Comment (bboure) [reply packages/graphql-language-service/src/utils/getVariablesJSONSchema.ts:151]: After double checking, there is no spec for `Date`.
Also, JSONSchema `date-time` is not fully compatible with the specs
e.g. offset is required

I might be able to come with a regex that follows the specs.

**Comment by acao** [comment]
ok, in that case, I will just open my PR against this one tonight, because that PR will take a while to get reviews before we can merge

**Comment by acao** [comment]
@bboure I think this is fine to merge now actually, it won't be released until we merge the Release Packages PR. Otherwise it would be hard for me to make a PR against this branch

**Comment by acao** [comment]
@bboure ok! this looks good to merge, thanks for all of this!
